From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 2 Feb 2025 06:08:38 +0100
Subject: [PATCH] Fix children copying when filling missing command redirects

When the redirect of some command wasn't found (e.g. because "commands.send-namespaced" was disabled in the spigot config), Paper would copy over the command like normal, but not iterate over the children and instead fully clone them
This fix now applies the same logic which is used for copying children normally

diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index ec1cced129ef42be65d7b2b622638bfae8bd895e..992d8d9e25e58ec0d5a0393d87d9d9b3a56dc21d 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -518,6 +518,12 @@ public class Commands {
     }
 
     private static <S> void fillUsableCommands(java.util.Collection<CommandNode<S>> children, CommandNode<S> current, S source, Map<CommandNode<S>, CommandNode<S>> output) { // Paper - Perf: Async command map building; pass copy of children
+        // CloudPlane start - support filling builders
+        fillUsableCommands(children, current::addChild, source, output);
+    }
+
+    private static <S> void fillUsableCommands(java.util.Collection<CommandNode<S>> children, Consumer<CommandNode<S>> current, S source, Map<CommandNode<S>, CommandNode<S>> output) { // Paper - Perf: Async command map building; pass copy of children
+        // CloudPlane end - support filling builders
         for (CommandNode<S> commandNode : children) {  // Paper - Perf: Async command map building; pass copy of children
             // Paper start - Brigadier API
             if (commandNode.clientNode != null) {
@@ -554,10 +560,10 @@ public class Commands {
                     if (redirect.getCommand() != null) {
                         builder.executes(redirect.getCommand());
                     }
-                    // Diff copied from LiteralCommand#createBuilder
-                    for (final CommandNode<S> child : redirect.getChildren()) {
-                        builder.then(child);
-                    }
+                    // CloudPlane start - don't blindly copy children over
+                    fillUsableCommands(commandNode.getRedirect().getChildren(),
+                            builder::then, source, output);
+                    // CloudPlane end - don't blindly copy children over
 
                     argumentBuilder = builder;
                 }
@@ -568,7 +574,7 @@ public class Commands {
 
                 CommandNode<S> commandNode1 = argumentBuilder.build();
                 output.put(commandNode, commandNode1);
-                current.addChild(commandNode1);
+                current.accept(commandNode1); // CloudPlane - support filling builders
                 if (!commandNode.getChildren().isEmpty()) {
                     fillUsableCommands(commandNode.getChildren(), commandNode1, source, output); // Paper - Perf: Async command map building; pass copy of children
                 }
